# Data Types and Structures

## Download Rmd Version

If you wish to engage with this course content via Rmd, then please click the link below to download the Rmd file.

[Download data_types.Rmd](rmarkdown/data_types.Rmd)

## Learning Objectives
- Understand and identify the five core data types used in R (Numeric, Integer, Complex, Character, Logical)
- Learn how to use functions like `typeof()`, `class()`, `mode()`, `length()`, and `attributes()` to examine the features of variables or objects in R
- Learn methods to address columns in data frames using the `$` operator or square brackets
- Practice defining variables and using R functions to profile their characteristics
- Understand and perform conversions between different data types in R

## Inbuilt Datasets
R has a number of datasets included for you to practice with. We can get a list of these by running the command ```data()```


In [2]:
data()

File: /var/folders/r7/wblx0jw96hz08nvjz9p3zsgr0000gp/T//RtmpcTNU8Z/RpackageIQR5eca5ebeecfa
Data sets in package ‘.’:



bp_dataset_session4     

code_session2           

plotting_arguments      

r_training_session_2_script

                        

r_training_session_3_script

                        

r_training_session_4    

worms                   



Data sets in package ‘datasets’:



AirPassengers           Monthly Airline Passenger Numbers 1949-1960

BJsales                 Sales Data with Leading Indicator

BJsales.lead (BJsales)

                        Sales Data with Leading Indicator

BOD                     Biochemical Oxygen Demand

CO2                     Carbon Dioxide Uptake in Grass Plants

ChickWeight             Weight versus age of chicks on different diets

DNase                   Elisa assay of DNase

EuStockMarkets          Daily Closing Prices of Major European Stock

                        Indices, 1991-1998

Formaldehyde            Determination of Form

This command won't produce any output but open a file in the scripts pane that lists the available datasets. 
In this session, we will be using the built-in R data sets stored in the variables ```iris``` & ```mtcars```. 
By having our data stored in a variable means we can easily use it with R functions and start to
manipulate or process it. Note that we are using the word data very loosely, it can refer to any form of 
information we want to store.

## Exploring datasets

Let's take a closer look at the `iris` data.

First, let's ask what type of thing `iris` is using the function ```class()``` for.


In [3]:
class(iris)

[1] "data.frame"


The output tells us that it is a data frame. A data frame is an example of an R object, and can be defined by certain properties. A data frame is comparable to a spreadsheet in MS Excel as it is a 2 dimensional object (i.e. has rows and columns). Data frames are very useful for storing data, especially if you are used to work with your data in tables. A typical data frame of experimental data contains individual observations in rows and variables in columns.
We can see the shape, or dimensions, of the data frame with the function ```dim()```:

In [4]:
dim(iris)

[1] 150   5


This tells us that our data frame, `iris`, has 150 rows and 5 columns.


To explore data frames, there are a number of relevant functions:

* ```head()``` - shows first 6 rows
* ```tail()``` - shows last 6 rows
* ```dim()``` - returns the dimensions of data frame (i.e. number of rows and number of columns)
* ```nrow()``` - number of rows
* ```ncol()```- number of columns
* ```str()``` - structure of data frame - name, type and preview of data in each column
* ```names()``` or ```colnames()``` - both show the names attribute for a data frame

# Subsetting Data

There are many occasions we want to "look" at some part of the data. Extracting a subset is known as slicing.
If we want to get a single value from the data frame, we can index a specific position using square brackets.
If you are familiar with matrices we index in the same way. For example, to get the element in the top left corner, i.e. in the first row and first column we run:

In [5]:
iris[1, 1]

[1] 5.1


We can use this principle to extract any entry of the matrix. For example the 30th row and 3rd column

In [6]:
iris[30,3]

[1] 1.6


An index like ``[30, 3]``` selects a single element of a data frame, but we can select larger sections as well. For example, we can select the first three columns of values for the first four rows like this:

In [7]:
iris[1:4, 1:3]

  Sepal.Length Sepal.Width Petal.Length
1          5.1         3.5          1.4
2          4.9         3.0          1.4
3          4.7         3.2          1.3
4          4.6         3.1          1.5


The slice ```1:4``` means, "Start at index 1 and go to index 4." The slice does not need to start at 1, e.g. the line below selects rows 5 through 10:

In [8]:
iris[5:10, 1:3]

   Sepal.Length Sepal.Width Petal.Length
5           5.0         3.6          1.4
6           5.4         3.9          1.7
7           4.6         3.4          1.4
8           5.0         3.4          1.5
9           4.4         2.9          1.4
10          4.9         3.1          1.5


We can use the function ```c()```, which stands for **c**ombine, to select non-contiguous entries:

In [9]:
iris[c(3, 8, 37, 56), c(1,3)]

   Sepal.Length Petal.Length
3           4.7          1.3
8           5.0          1.5
37          5.5          1.3
56          5.7          4.5


We also don't have to provide a slice for either the rows or the columns. If we don't include a slice for the rows, R returns all the rows; if we don't include a slice for the columns, R returns all the columns. If we don't provide a slice for either rows or columns, e.g. `iris[, ]`, R returns the full data frame.


In [10]:
iris[5, ]

  Sepal.Length Sepal.Width Petal.Length Petal.Width Species
5            5         3.6          1.4         0.2  setosa


In [11]:
iris[, 2]

  [1] 3.5 3.0 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 3.7 3.4 3.0 3.0 4.0 4.4 3.9 3.5
 [19] 3.8 3.8 3.4 3.7 3.6 3.3 3.4 3.0 3.4 3.5 3.4 3.2 3.1 3.4 4.1 4.2 3.1 3.2
 [37] 3.5 3.6 3.0 3.4 3.5 2.3 3.2 3.5 3.8 3.0 3.8 3.2 3.7 3.3 3.2 3.2 3.1 2.3
 [55] 2.8 2.8 3.3 2.4 2.9 2.7 2.0 3.0 2.2 2.9 2.9 3.1 3.0 2.7 2.2 2.5 3.2 2.8
 [73] 2.5 2.8 2.9 3.0 2.8 3.0 2.9 2.6 2.4 2.4 2.7 2.7 3.0 3.4 3.1 2.3 3.0 2.5
 [91] 2.6 3.0 2.6 2.3 2.7 3.0 2.9 2.9 2.5 2.8 3.3 2.7 3.0 2.9 3.0 3.0 2.5 2.9
[109] 2.5 3.6 3.2 2.7 3.0 2.5 2.8 3.2 3.0 3.8 2.6 2.2 3.2 2.8 2.8 2.7 3.3 3.2
[127] 2.8 3.0 2.8 3.0 2.8 3.8 2.8 2.8 2.6 3.0 3.4 3.1 3.0 3.1 3.1 3.1 2.7 3.2
[145] 3.3 3.0 2.5 3.0 3.4 3.0


# Addressing Columns by name 
Columns can also be addressed by name, with either the ```$``` operator (i.e. ```iris$Petal.Length```) or square brackets (ie. ```iris[,'Petal.Length']```).

# Data Types 

One key feature of a data frame is each column is classed as a specific data type. 

*Data types*, or modes, define what the values are and how they can be used. 

There are __*5 core data types*:__

* Numeric - all real numbers e.g. 7.5
* Integer - e.g. 2
* Complex - numbers with real and imaginary parts e.g. 1+4i
* Character - consists of letters or numbers or symbols or a combination of these e.g. "a", "f5", "datatypes", "Learning R is fun"
* Logical - only takes TRUE or FALSE.


These data types are also used to characterise other one dimensional R objects such as individual values or vectors 
(more on these later). R provides a number of functions to examine the features of variables or objects. 

Some examples:

* ```typeof()``` - what is the object’s data type (on the data storage level ("what the computer sees"))?
* ```class()``` - what is the object's data type (on the abstract type level("what R sees"))?
* ```mode()``` - what is the object's data type (on the data storage level ("what the computer sees"))?
* ```length()```- how long is it? 
* ```attributes()``` - does it have any metadata? 
* ```str``` - display the internal structure of an object.
* ```is.numeric()```, ```is.character()```, ```is.complex()```, ```is.logical()``` - returns TRUE when an object is the datatype queried, FALSE if not.

Let's define some variables we can use R to profile the charateristics of

In [12]:
x <- "dataset"
typeof(x)

[1] "character"


In [13]:
y <- 1:10
y

 [1]  1  2  3  4  5  6  7  8  9 10


In [14]:
typeof(y)

[1] "integer"


In [15]:
class(y)

[1] "integer"


In [16]:
length(y)

[1] 10


# Converting Between Datatypes

It can be critical that R has correctly assigned the right data type to your variable. If it has not you may run into 
errors when processing it. You therefore may want to convert between different data types. This can be done with the 
series of functions ```as.numeric```, ```as.character``` etc. 

For example, let's convert our numeric variable `y` into a character. 

In [17]:
y <- as.character(y)
typeof(y)

[1] "character"


## Activity: Determining Data Yypes

Create a variable with the numbers 9,2,200, and 14.
What class do you predict this variable to be? Use an R function to confirm your answer.
If it is not the data type you expected, can you force R to convert it to an integer?
Divide each element of the variable in half. Does this change the type of variable?


# Data Structures

R has a number of inbuilt structures that can be used to store datasets. We have encountered one of these already - the data.frame. Other include: 

* strings
* vectors 
* data frames
* matrices
* arrays
* lists
* factors

## Strings

A string is a run of characters. e.g. "hello". or "a189jde2mjo". They are enclosed in quotes.

## Vectors

A vector is the most common and basic data structure in R. They are an ordered collection of basic data types of 
a given length. They are one-dimensional. We can think of each column of a data frame as a vector. 

The concatenate or combine ```c()``` function will explicitly __construct__ a vector. 



In [18]:
v_num <- c(9,20,12)
v_log <- c(TRUE, FALSE, FALSE, TRUE)

As we create these vectors you should see them listed in your environment pane. We can then call the name of the varible to see what the value is at that point in time. 

In [19]:
v_num 

[1]  9 20 12


In [20]:
v_log

[1]  TRUE FALSE FALSE  TRUE


The function ```c()``` can also be used to __add elements__ to a vector.

In [21]:
v_name <- c("Sarah", "Tracy", "Jon")
v_name <- c(v_name, "Annette")
v_name

[1] "Sarah"   "Tracy"   "Jon"     "Annette"


If we want to create a series of numbers, like 1 to 140, we can bypass the `c()` constructor and just write

In [22]:
(v_int <- 1:4)

[1] 1 2 3 4


When we call a vector *atomic*, we mean that the vector only holds data of a single data type.

## Data Frame

A data frame is a two dimensional structure consisting of rows and columns. 

We can create a data frames using the function `data.frame()`. We will use the predefined vector `letter` 
to get the first ten letters in the alphabet.


In [23]:
dat <- data.frame(id = letters[1:10], x = 1:10, y = 11:20)
dat

   id  x  y
1   a  1 11
2   b  2 12
3   c  3 13
4   d  4 14
5   e  5 15
6   f  6 16
7   g  7 17
8   h  8 18
9   i  9 19
10  j 10 20


Columns are variables, and each column will have a specified data type, which all entries must adhere to. 
Rows are observations. Rows and columns will have `rownames` and `colnames` that you can use to extract 
specific rows or columns respectively. 

If you read in a table of data from a file, it will typically be represented by a data.frame.


### Functions to explore data frames

* ```head()``` - shows first 6 rows
* ```tail()``` - shows last 6 rows
* ```dim()``` - returns the dimensions of data frame (i.e. number of rows and number of columns)
* ```nrow()``` - number of rows
* ```ncol()```- number of columns
* ```str()``` - structure of data frame - name, type and preview of data in each column
* ```names()``` or ```colnames()``` - both show the names attribute for a data frame


## Matrices

A matrix is another two dimensional object, but it differs to a data.frame as all columns/entries must be of the same type. 
It is more efficient memory-wise than a data.frame, but can-not be used as a substitute to all data.frames.

We can construct a matrix as follows:

In [24]:
m<-matrix(1:6, nrow=2, ncol=3)
m

     [,1] [,2] [,3]
[1,]    1    3    5
[2,]    2    4    6


When creating a __matrix__, it is important to remember that matrices __*are filled column-wise*__. If that is not what you want, you can use the ```byrow``` argument (a logical: can be ```TRUE``` or ```FALSE```) to specify how the matrix is filled.

We can confirm the data type of a matrix with the function `class`.

In [25]:
class(m)

[1] "matrix" "array" 


## Arrays

Arrays are n-dimensional storage structures. A one-dimensional array is a vector, a two-dimensional array is a matrix. 

We will not be using this type of object, but it is included for completeness.

## Lists

A list in R is a collection of objects and elements, which themselves can be a heterogeneous mix of other objects 
including vectors, matrices, data.frames, functions, strings, numbers. They tend to be used to collate different 
data types connected in some way. 

We can create a list of the vectors and data.frames we have already constructed as follows:

In [26]:
l <- list(v_log, v_num, iris)
l

[[1]]
[1]  TRUE FALSE FALSE  TRUE

[[2]]
[1]  9 20 12

[[3]]
    Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
1            5.1         3.5          1.4         0.2     setosa
2            4.9         3.0          1.4         0.2     setosa
3            4.7         3.2          1.3         0.2     setosa
4            4.6         3.1          1.5         0.2     setosa
5            5.0         3.6          1.4         0.2     setosa
6            5.4         3.9          1.7         0.4     setosa
7            4.6         3.4          1.4         0.3     setosa
8            5.0         3.4          1.5         0.2     setosa
9            4.4         2.9          1.4         0.2     setosa
10           4.9         3.1          1.5         0.1     setosa
11           5.4         3.7          1.5         0.2     setosa
12           4.8         3.4          1.6         0.2     setosa
13           4.8         3.0          1.4         0.1     setosa
14           4.3         3.0 

## Factors

Sometimes considered as a data type, as confusingly it is a possible response to the function `type`, and therefore
a valid option for columns in a data frame. A factor is designed for categorical variables. They have a finite number of 
"levels", which are the options that any element of that vector can take. They are actually stored as integers 
which can make them quite powerful for subsetting. 

It is very easy for character vectors to be inappropriately stored as factors and vice versa. In fact, R's default 
when loading data is to store a string as a factor. Conversely if we define a vector, we have to explicitly convert 
it to a vector.

For example, if we define a vector of months, the default is to class it as a vector of characters. We have to actively 
coerce it into a factor. 

In [27]:
a      <- c("March","February","February","November","February","March","March","March","February","November")
class(a)

[1] "character"


In [28]:
fact <- as.factor(a)
class(a)

[1] "character"


## Activity: Data Exploration

Use these functions to explore the mtcars dataset.

* How large is the dataset?
* What type is the object?
* What value is in the 6th row of the 4th column?

## Summary Quiz

In [1]:
library(jsonlite)
library(IRdisplay)

show_quiz_from_json <- function(path) {
  quiz_data <- fromJSON(path, simplifyVector = FALSE)
  
  html <- '
  <style>
  .quiz-question {
      background-color: #6c63ff;
      color: white;
      padding: 12px;
      border-radius: 10px;
      font-weight: bold;
      font-size: 1.2em;
      margin-bottom: 10px;
  }

  .quiz-form {
      margin-bottom: 20px;
  }

  .quiz-answer {
      display: block;
      background-color: #f2f2f2;
      border: none;
      border-radius: 10px;
      padding: 10px;
      margin: 5px 0;
      font-size: 1em;
      cursor: pointer;
      text-align: left;
      transition: background-color 0.3s;
      width: 100%;
  }

  .quiz-answer:hover {
      background-color: #e0e0e0;
  }

  .correct {
      background-color: #4CAF50 !important;
      color: white !important;
      border: none;
  }

  .incorrect {
      background-color: #D32F2F !important;
      color: white !important;
      border: none;
  }

  .feedback {
      margin-top: 10px;
      font-weight: bold;
      font-size: 1em;
  }
  </style>

  <script>
  function handleAnswer(qid, aid, feedback, isCorrect) {
      let buttons = document.querySelectorAll(".answer-" + qid);
      buttons.forEach(btn => {
          btn.classList.remove("correct", "incorrect");
      });

      let selected = document.getElementById(aid);
      selected.classList.add(isCorrect ? "correct" : "incorrect");

      let feedbackBox = document.getElementById("feedback_" + qid);
      feedbackBox.innerHTML = feedback;
      feedbackBox.style.color = isCorrect ? "green" : "red";
  }
  </script>
  '

  for (i in seq_along(quiz_data)) {
    question <- quiz_data[[i]]
    qid <- as.character(i)
    html <- paste0(html, sprintf('<div class="quiz-question">%s</div><form class="quiz-form">', question$question))
    
    for (j in seq_along(question$answers)) {
      answer <- question$answers[[j]]
      aid <- sprintf("q%s_a%s", i, j)
      feedback <- gsub("'", "\\'", answer$feedback)  # escape single quotes
      correct <- tolower(substr(answer$feedback, 1, 7)) == "correct"
      correct_js <- tolower(as.character(correct))  # "true" or "false"

      html <- paste0(html, sprintf(
        '<button type="button" class="quiz-answer answer-%s" id="%s"
        onclick="handleAnswer(\'%s\', \'%s\', \'%s\', %s)">
        %s</button>',
        qid, aid, qid, aid, feedback, correct_js, answer$answer
      ))
    }
    
    html <- paste0(html, sprintf('<div class="feedback" id="feedback_%s"></div></form><hr>', qid))
  }

  display_html(html)
}

# Call the function to display quiz interactively:
show_quiz_from_json("questions/summary_basic_commands.json")

What does a relative file path NOT start with? 
 A drive letter (e.g., C:) 
 A forward slash (/) 
 A dot (.) 
 A folder name Which operator is used to assign a value to a variable in R? 
 == 
 -> 
 = 
 <- What is the primary use of the console pane in RStudio? 
 To edit scripts and save them for later use 
 To browse files and manage projects 
 To execute R commands and display results 
 To visualize data using plots